File: mvp/ts/src/pipeline.ts
import { MqttSource } from "./adapters/source/mqtt";
import { HttpSink } from "./adapters/sink/http_push";
import { parseDevicePayload } from "./parse";
import { toExternalCamel } from "./util/casing";

export class Pipeline {
  source: any; sink: any;
  constructor(source:any, sink:any){ this.source=source; this.sink=sink; }
  async runOnce(){
    for (const raw of this.source.stream()){
      const reading = parseDevicePayload(raw);
      const payload = toExternalCamel(reading);
      await this.sink.send(payload);
    }
  }
}

export function buildFromConfig(cfg:any){
  return new Pipeline(new MqttSource(cfg.source), new HttpSink(cfg.sink));
}

File: mvp/python/app/pipeline.py
from mvp.python.adapters.source.mqtt import MqttSource
from mvp.python.adapters.sink.http_push import HttpSink
from mvp.python.app.parse import parse_device_payload

class Pipeline:
    def __init__(self, source, sink):
        self.source = source; self.sink = sink
    def run_once(self):
        for raw in self.source.stream():
            reading = parse_device_payload(raw)
            self.sink.send(reading.model_dump(by_alias=False))

def build_from_config(cfg):
    return Pipeline(MqttSource(cfg['source']), HttpSink(cfg['sink']))
